src/sync/list.rs

```diff
@@ -1,125 +1,175 @@
-    pub fn insert<'scope>(&self, data: T, scope: &'scope Scope) -> Ptr<'scope, Entry<T>> {
-        let mut new = Owned::new(Entry {
-            data: data,
-            next: Atomic::null(),
-        });
-        let mut head = self.head.load(Relaxed, scope);
+    pub fn insert<'scope>(
+        &'scope self,
+        to: &'scope Atomic<Node<T>>,
+        data: T,
+        scope: &'scope Scope,
+    ) -> Ptr<'scope, Node<T>> {
+        let mut cur = Owned::new(Node::new(data));
+        let mut next = to.load(Relaxed, scope);
 
         loop {
-            new.next.store(head, Relaxed);
-            match self.head.compare_and_set_weak_owned(head, new, Release, scope) {
-                Ok(n) => return n,
-                Err((h, n)) => {
-                    head = h;
-                    new = n;
+            cur.0.next.store(next, Relaxed);
+            match to.compare_and_set_weak_owned(next, cur, Release, scope) {
+                Ok(cur) => return cur,
+                Err((n, c)) => {
+                    next = n;
+                    cur = c;
                 }
             }
         }
     }
 
+    pub fn insert_head<'scope>(
+        &'scope self,
+        data: T,
+        scope: &'scope Scope,
+    ) -> Ptr<'scope, Node<T>>
+where {
+        self.insert(&self.head, data, scope)
+    }
+
     /// Returns an iterator over all data.
     ///
     /// Every datum that is inserted at the moment this function is called and persists at least
-    /// until the end of iteration will be returned. Since this iterator traverses a lock-free linked
-    /// list that may be concurrently modified, some additional caveats apply:
+    /// until the end of iteration will be returned. Since this iterator traverses a lock-free
+    /// linked list that may be concurrently modified, some additional caveats apply:
     ///
     /// 1. If a new datum is inserted during iteration, it may or may not be returned.
     /// 2. If a datum is deleted during iteration, it may or may not be returned.
-    /// 3. Any datum that gets returned may be returned multiple times.
-    pub fn iter<'scope>(&'scope self, scope: &'scope Scope) -> Iter<'scope, T> {
-        let head = &self.head;
-        let pred = head;
+    /// 3. It may not return all data if a concurrent thread continues to iterate the same list.
+    pub fn iter<'scope>(&'scope self, scope: &'scope Scope) -> Iter<'scope, T>
+where {
+        let pred = &self.head;
         let curr = pred.load(Acquire, scope);
-        Iter { scope, head, pred, curr }
+        Iter { scope, pred, curr }
     }
 }
 
-impl<'scope, T> Iterator for Iter<'scope, T> {
-    type Item = &'scope T;
+impl<T> Drop for List<T> {
+    fn drop(&mut self) {
+        unsafe {
+            unprotected(|scope| {
+                let mut curr = self.head.load(Relaxed, scope);
+                while let Some(c) = curr.as_ref() {
+                    let succ = c.0.next.load(Relaxed, scope);
+                    scope.defer_free(curr);
```

Comment:
In destructors you don't have to defer destruction - you can destroy immediately.